### PR TITLE
security(logging): fix aes implementation in audit logging

### PR DESCRIPTION
**Pull request by joshua-goldstein**
## Background

Audit logging is an enterprise feature with an option for encryption using AES.  An AES cipher has three ingredients: a private key, an iv (or nonce), and plaintext.  The nonce is public but it must be unique.

## Problem

Currently we use a [16]byte nonce.  The first 12 bytes come from a baseIv which is initialized when an audit log is created.  The last 4 bytes come from the length of the log line being encrypted.  This is problematic because two log lines will often have the same length, so due to these collisions we are reusing the same nonce many times.  AES requires a unique nonce and without a unique nonce it is trivial to break.  Many thanks to @HakuPiku for pointing this out to us.

## Solution

Use crypto/rand to generate a unique nonce every time we encrypt a log line.  Previously we used what was essentially a 16 byte header (12 byte baseIv and 4 byte length of line).  We now use a [20] byte header, consisting of a unique nonce (16 bytes) and length of line (4 bytes).

## Remarks

- For audit logging, _x/log_writer.go_ controls all the audit logging (i.e., writing) while  _ee/audit/run_ee.go_ does all the decrypting.  There are other encryption/decryption tools (in testutil, e.g.) that are used for other purposes (e.g., exports).
- Customers may have old audit logs that will have to be decrypted using the deprecated method.  Now the decrypt method will determine whether the audit log is old or new and use the appropriate scheme.
- This PR also adds several integration tests to check if we can appropriately encrypt and decrypt audit logs (previously we had no integration tests for this, only a small unit test)
- We also add tests to check if an audit log in the old format can be decrypted correctly.  This is the reason there is an `.enc` file checked into the PR.
- Finally we add an extra unit test to verify that encrypted audit logs can be truncated and still decrypted successfully (up to the line that was truncated)

## References
- https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation
- https://pkg.go.dev/crypto/cipher#example-NewCTR

### CODE REVIEW

**Comment by HakuPiku** [comment]
Any updates on when this will be merged and a security advisory created?

File: ee/audit/run_ee.go [hunk 1]
```diff
@@ -107,37 +107,116 @@ func run() error {
...
 
-	for {
-		// if its the end of data. finish decrypting
-		if iterator >= stat.Size() {
-			break
-		}
-		x.Check2(file.ReadAt(iv[12:], iterator))
-		iterator = iterator + 4
+	// [12]byte baseIV + [4]byte len(x.VerificationText_Deprecated) + [11]byte x.VerificationText_Deprecated
+	decryptHeader_Deprecated := func() ([]byte, int64, error) {
```
Comment (mangalaman93) [inline ee/audit/run_ee.go:133]: no underscore in function names

File: x/log_writer.go [hunk 2]
```diff
@@ -41,7 +41,9 @@ const (
 	backupTimeFormat = "2006-01-02T15-04-05.000"
 	bufferSize       = 256 * 1024
 	flushInterval    = 10 * time.Second
-	VerificationText = "Hello World"
+	//  old logs before https://github.com/dgraph-io/dgraph/pull/8323 contain deprecated verification text in header
+	VerificationText_Deprecated = "Hello World"
+	VerificationText            = "dlroW olloH"
```
Comment (mangalaman93) [inline x/log_writer.go:46]: is it really supposed to be olloH? or olleH?

File: ee/audit/run_ee.go [hunk 3]
```diff
@@ -107,37 +107,116 @@ func run() error {
...
+		var iterator int64 = 0
+		iv := make([]byte, aes.BlockSize)
+		x.Check2(file.ReadAt(iv, iterator))
+		iterator = iterator + aes.BlockSize
+
+		t := make([]byte, len(x.VerificationText_Deprecated))
+		x.Check2(file.ReadAt(t, iterator))
+		iterator = iterator + int64(len(x.VerificationText_Deprecated))
```
Comment (mangalaman93) [inline ee/audit/run_ee.go:141]: same for variable name, no underscore

File: ee/audit/run_ee.go [hunk 4]
```diff
@@ -107,37 +107,116 @@ func run() error {
...
-		x.Check2(file.ReadAt(iv[12:], iterator))
-		iterator = iterator + 4
+	// [12]byte baseIV + [4]byte len(x.VerificationText_Deprecated) + [11]byte x.VerificationText_Deprecated
+	decryptHeader_Deprecated := func() ([]byte, int64, error) {
+		var iterator int64 = 0
+		iv := make([]byte, aes.BlockSize)
+		x.Check2(file.ReadAt(iv, iterator))
+		iterator = iterator + aes.BlockSize
+
+		t := make([]byte, len(x.VerificationText_Deprecated))
```
Comment (mangalaman93) [inline ee/audit/run_ee.go:139]: better name than `t`

File: ee/audit/run_ee.go [hunk 5]
```diff
@@ -107,37 +107,116 @@ func run() error {
...
-		if iterator >= stat.Size() {
-			break
-		}
-		x.Check2(file.ReadAt(iv[12:], iterator))
-		iterator = iterator + 4
+	// [12]byte baseIV + [4]byte len(x.VerificationText_Deprecated) + [11]byte x.VerificationText_Deprecated
+	decryptHeader_Deprecated := func() ([]byte, int64, error) {
+		var iterator int64 = 0
+		iv := make([]byte, aes.BlockSize)
+		x.Check2(file.ReadAt(iv, iterator))
```
Comment (mangalaman93) [inline ee/audit/run_ee.go:136]: does it make sense to crash when we can't read from the file?

File: ee/audit/run_ee.go [hunk 6]
```diff
@@ -107,37 +107,116 @@ func run() error {
...
+		iterator = iterator + int64(len(x.VerificationText))
 
-	stream := cipher.NewCTR(block, iv)
-	stream.XORKeyStream(t, t)
-	if string(t) != x.VerificationText {
-		return errors.New("invalid encryption key provided. Please check your encryption key")
+		text := make([]byte, len(x.VerificationText))
+		stream := cipher.NewCTR(block, iv)
+		stream.XORKeyStream(text, t)
+		if string(text) != x.VerificationText {
```
Comment (mangalaman93) [inline ee/audit/run_ee.go:126]: can we definer the VerificationText of type []byte and use bytes.Compare instead?

File: systest/audit_encrypted/audit_test.go [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+	out, err := decrypt.CombinedOutput()
+	if err != nil {
+		fmt.Printf("Error %v\n", err)
+		fmt.Printf("Output %v\n", string(out))
+		t.Fatal(err)
+	}
+
+	verifyLogs(t, fmt.Sprintf("audit_dir/za/zero_audit_0_%s.log", nId), msgs)
+}
+func TestAlphaAuditEncrypted(t *testing.T) {
```
Comment (mangalaman93) [inline systest/audit_encrypted/audit_test.go:83]: leave a newline before this

File: systest/audit_encrypted/audit_test.go [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+		fmt.Printf("Output %v\n", string(out))
+		t.Fatal(err)
+	}
+
+	verifyLogs(t, fmt.Sprintf("audit_dir/aa/alpha_audit_1_%s.log", nId), msgs)
+}
+
+// deprecated logs generated with dgraph v22.0.2
+func TestZeroAuditDecryptDeprecated(t *testing.T) {
+	if len(zeroMsgs) == 0 {
```
Comment (mangalaman93) [inline systest/audit_encrypted/audit_test.go:166]: I think this is not necessary, the test could run in either order. Each test should be independent.

File: systest/audit_encrypted/audit_test.go [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+
+	var args []string
+	args = append(args, "audit", "decrypt",
+		"--encryption_key_file=../../ee/enc/test-fixtures/enc-key",
+		"--in", "audit_dir_deprecated/za/zero_audit_0_1.log.enc",
+		"--out", "audit_dir_deprecated/za/zero_audit_0_1.log")
+
+	decrypt := exec.Command(testutil.DgraphBinaryPath(), args...)
+
+	fmt.Println("Running: ", decrypt.Args)
```
Comment (mangalaman93) [inline systest/audit_encrypted/audit_test.go:179]: use t.Log instead

File: systest/audit_encrypted/docker-compose.yml [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+    working_dir: /data/alpha1
+    ports:
+      - "8080"
+      - "9080"
+    volumes:
+      - type: bind
+        source: $GOPATH/bin
+        target: /gobin
+        read_only: true
+      - type: bind
```
Comment (mangalaman93) [inline systest/audit_encrypted/docker-compose.yml:14]: use a volume instead

File: x/log_writer.go [hunk 11]
```diff
@@ -236,6 +306,109 @@ func (l *LogWriter) open() error {
...
+			}
+		}
+	}
+
+	l.writer = bufio.NewWriterSize(l.file, bufferSize)
+	l.size = size()
+	return nil
+}
+
+func (l *LogWriter) open_deprecated() error {
```
Comment (mangalaman93) [inline x/log_writer.go:395]: Have we added these functions?

**Review by mangalaman93 — CHANGES_REQUESTED** [review]
1. Do we need the .gitkeep files?

File: x/log_writer.go (hunk 11 above)
Comment (joshua-goldstein) [reply x/log_writer.go:395]: this is the code that existed there previously - I just appended _deprecated to the function name.  If you feel it is not necessary to keep this function in the code I can remove it

File: x/log_writer.go (hunk 2 above)
Comment (joshua-goldstein) [reply x/log_writer.go:46]: Good catch :)  all we need is a string of length 11 that is different from "Hello World".  I can change it but it wouldn't make a difference

File: systest/audit_encrypted/docker-compose.yml (hunk 10 above)
Comment (joshua-goldstein) [reply systest/audit_encrypted/docker-compose.yml:14]: > 1. Do we need the .gitkeep files?

I believe so, since we want to keep the audit_dir directory structure intact.  See e.g. systest/audit.  I copied over the directories from there.

File: x/log_writer.go (hunk 11 above)
Comment (mangalaman93) [reply x/log_writer.go:395]: we didn't add the replacement code here then? and I think we can remove the code that is not used, we can always find in the history.

File: systest/audit_encrypted/docker-compose.yml (hunk 10 above)
Comment (mangalaman93) [reply systest/audit_encrypted/docker-compose.yml:14]: You can use a volume (not volume mount) and probably you wouldn't need those folders.

File: ee/audit/run_ee.go [hunk 12]
```diff
@@ -107,37 +107,116 @@ func run() error {
...
+	decryptHeader := func() ([]byte, int64, error) {
+		var iterator int64 = 0
+		iv := make([]byte, aes.BlockSize)
+		x.Check2(file.ReadAt(iv, iterator))     // get first iv
+		iterator = iterator + aes.BlockSize + 4 // length of verification text encoded in uint32
 
-	t := make([]byte, len(x.VerificationText))
-	x.Check2(file.ReadAt(t, iterator))
-	iterator = iterator + int64(len(x.VerificationText))
+		t := make([]byte, len(x.VerificationText))
```
[the remark this answers is not in the document]
Comment (joshua-goldstein) [reply ee/audit/run_ee.go:119]: done

File: ee/audit/run_ee.go (hunk 1 above)
Comment (joshua-goldstein) [reply ee/audit/run_ee.go:133]: done

File: ee/audit/run_ee.go (hunk 3 above)
Comment (joshua-goldstein) [reply ee/audit/run_ee.go:141]: done

File: ee/audit/run_ee.go (hunk 4 above)
Comment (joshua-goldstein) [reply ee/audit/run_ee.go:139]: done

File: systest/audit_encrypted/audit_test.go (hunk 7 above)
Comment (joshua-goldstein) [reply systest/audit_encrypted/audit_test.go:83]: done

File: systest/audit_encrypted/audit_test.go (hunk 9 above)
Comment (joshua-goldstein) [reply systest/audit_encrypted/audit_test.go:179]: done

File: x/log_writer.go (hunk 11 above)
Comment (joshua-goldstein) [reply x/log_writer.go]: sounds reasonable to me - I remove the unused code and kept only the functions necessary to decrypt old audit logs

File: ee/audit/run_ee.go (hunk 5 above)
Comment (joshua-goldstein) [reply ee/audit/run_ee.go]: Good catch...  this is indeed an issue because we may try using multiple decrypt methods and don't want to crash and should handle the error more gracefully.

File: systest/audit_encrypted/docker-compose.yml (hunk 10 above)
Comment (joshua-goldstein) [reply systest/audit_encrypted/docker-compose.yml:14]: done

File: systest/audit_encrypted/audit_test.go (hunk 8 above)
Comment (joshua-goldstein) [reply systest/audit_encrypted/audit_test.go]: done

File: systest/audit_encrypted/docker-compose.yml (hunk 10 above)
Comment (joshua-goldstein) [reply systest/audit_encrypted/docker-compose.yml:14]: moving back to a bind mount - as we discussed using a mount volume created other issues and is not necessary here since we are not persisting data beyond the life of the test.  The only "issue" is that .gitkeep directories are aesthetically unpleasing, but we already do this elsewhere in the repo (e.g. in systest/audit package) so we can revisit this later in the context of making a broader change (to other test packages as well)

File: ee/audit/run_ee.go (hunk 6 above)
Comment (joshua-goldstein) [reply ee/audit/run_ee.go:132]: Currently we declare a string constant in `x/log_writer.go` .  I think this is reasonable since this should be a constant, and we can't declare an array/slice to be a constant.  We could cast the x.VerificationText as a byte slice and compare the bytes but I don't think this makes a difference at all

File: ee/audit/run_ee.go [hunk 13]
```diff
@@ -107,37 +107,161 @@ func run() error {
...
+	// #################################################################
+	decryptBody := func() {
+		for {
+			count++
+			// if its the end of data. finish decrypting
+			if iterator >= stat.Size() {
+				break
+			}
+			n, err := file.ReadAt(iv, iterator)
+			if err != nil {
```
Comment (mangalaman93) [inline ee/audit/run_ee.go:193]: shouldn't we break the loop or something?

File: ee/audit/run_ee.go [hunk 14]
```diff
@@ -107,37 +107,161 @@ func run() error {
...
+	// #####   [16]byte IV + [4]byte uint32(len(p)) + [:]byte p    #####
+	// #################################################################
+	decryptBody := func() {
+		for {
+			count++
+			// if its the end of data. finish decrypting
+			if iterator >= stat.Size() {
+				break
+			}
+			n, err := file.ReadAt(iv, iterator)
```
Comment (mangalaman93) [inline ee/audit/run_ee.go:192]: we should also check whether n is same as iterator probably here. I think there is a better way to write this whole code. You could take some inspiration from this code maybe https://dave.cheney.net/2019/01/27/eliminate-error-handling-by-eliminating-errors

File: ee/audit/run_ee.go [hunk 15]
```diff
@@ -107,37 +107,161 @@ func run() error {
...
+			if err != nil {
+				glog.Warningf("received %v while decrypting audit log\n", err)
+				glog.Warningf("read %v bytes, expected %v\n", n, len(iv))
+			}
+			iterator = iterator + 16
+			length := make([]byte, 4)
+			n, err = file.ReadAt(length, iterator)
```
Comment (mangalaman93) [inline ee/audit/run_ee.go:199]: same here

**Review by mangalaman93 — CHANGES_REQUESTED** [review]
I think we should consider writing more robust functions and test. This needs more work.

File: ee/audit/run_ee.go [hunk 16]
```diff
@@ -107,37 +107,161 @@ func run() error {
...
 
-	for {
-		// if its the end of data. finish decrypting
-		if iterator >= stat.Size() {
-			break
+	// [12]byte baseIV + [4]byte len(x.VerificationTextDeprecated) + [11]byte x.VerificationTextDeprecated
+	decryptHeaderDeprecated := func() ([]byte, int64, error) {
+		var iterator int64 = 0
+
+		iv := make([]byte, aes.BlockSize)
```
[the remark this answers is not in the document]
Comment (joshua-goldstein) [reply ee/audit/run_ee.go:142]: The length of the blocksize is 16 in both cases.  Previously we were using a baseIV (12 bytes) appended with the length of the cipher text (4 bytes, uint32) as the IV.  Now we randomly generate a 16 byte IV each time, followed by the length of the cipher text (4 bytes, uint32).  This is the point of the PR, because what we were doing before (baseIV + len(ciphertext)) introduces the possibility of collisions.

File: ee/audit/run_ee.go [hunk 17]
```diff
@@ -107,37 +107,161 @@ func run() error {
...
 		return nil
 	}
-	var iterator int64 = 0
 
-	iv := make([]byte, aes.BlockSize)
-	x.Check2(file.ReadAt(iv, iterator))
-	iterator = iterator + aes.BlockSize
+	// decrypt header in audit log to verify encryption key
+	// [16]byte IV + [4]byte len(x.VerificationText) + [11]byte x.VerificationText
+	decryptHeader := func() ([]byte, int64, error) {
```
[the remark this answers is not in the document]
Comment (joshua-goldstein) [reply ee/audit/run_ee.go:113]: What is the purpose of making them the same function?  For the updated audit logs the header looks like
```
[16]byte IV + [4]byte len(x.VerificationText) + [11]byte x.VerificationText
```

whereas for the old audit logs the header is
```
[12]byte baseIV + [4]byte len(x.VerificationTextDeprecated) + [11]byte x.VerificationTextDeprecated
```

and therefore they need to be parsed differently.  It is possible that we would have to try both functions to decrypt the header, so it would be very confusing to try to put the parsing logic of both inside one function.

File: ee/audit/run_ee.go [hunk 18]
```diff
@@ -107,37 +107,161 @@ func run() error {
...
+		var iterator int64 = 0
+		iv := make([]byte, aes.BlockSize)
+		_, err := file.ReadAt(iv, iterator) // get first iv
+		if err != nil {
+			return nil, 0, err
```
[the remark this answers is not in the document]
Comment (joshua-goldstein) [reply ee/audit/run_ee.go:118]: what kind of information?  this error is propagated up, and if we are also unable to decrypt the audit log header with `decryptHeaderDeprecated` then we return `errors.New("invalid encryption key provided. Please check your encryption key")` to the caller, the run() function

File: ee/audit/run_ee.go [hunk 19]
```diff
@@ -107,37 +107,161 @@ func run() error {
...
+		}
+		iterator = iterator + aes.BlockSize + 4 // length of verification text encoded in uint32
 
-	t := make([]byte, len(x.VerificationText))
-	x.Check2(file.ReadAt(t, iterator))
-	iterator = iterator + int64(len(x.VerificationText))
+		ct := make([]byte, len(x.VerificationText))
+		_, err = file.ReadAt(ct, iterator)
+		if err != nil {
+			return nil, 0, err
```
[the remark this answers is not in the document]
Comment (joshua-goldstein) [reply ee/audit/run_ee.go:125]: again, what kind of context?  this will be propagated up and return the invalid encryption key error to the user

File: ee/audit/run_ee.go [hunk 20]
```diff
@@ -107,37 +107,161 @@ func run() error {
...
+		}
+		iterator = iterator + int64(len(x.VerificationText))
 
-	stream := cipher.NewCTR(block, iv)
-	stream.XORKeyStream(t, t)
```
[the remark this answers is not in the document]
Comment (joshua-goldstein) [reply ee/audit/run_ee.go:121]: done

File: ee/audit/run_ee.go [hunk 21]
```diff
@@ -107,37 +107,161 @@ func run() error {
...
-		iterator = iterator + 4
+		iterator = iterator + aes.BlockSize
 
-		content := make([]byte, binary.BigEndian.Uint32(iv[12:]))
-		x.Check2(file.ReadAt(content, iterator))
-		iterator = iterator + int64(binary.BigEndian.Uint32(iv[12:]))
+		ct := make([]byte, len(x.VerificationTextDeprecated))
+		_, err = file.ReadAt(ct, iterator)
+		if err != nil {
+			return nil, 0, err
```
[the remark this answers is not in the document]
Comment (joshua-goldstein) [reply ee/audit/run_ee.go:151]: again, what kind of context? this will be propagated up and return the invalid encryption key error to the user

File: ee/audit/run_ee.go (hunk 14 above)
Comment (joshua-goldstein) [reply ee/audit/run_ee.go:190]: n is not the same as the iterator; it is the increment by which we need to increment the iterator.  I made a few changes to make this more clear

File: ee/audit/run_ee.go (hunk 13 above)
Comment (joshua-goldstein) [reply ee/audit/run_ee.go:191]: yes, added break statement here

File: ee/audit/run_ee.go [hunk 22]
```diff
@@ -107,37 +107,161 @@ func run() error {
...
+	// #####   [4]byte uint32(len(p)) + [:]byte p         #####
+	// ########################################################
+	decryptBodyDeprecated := func() {
+		for {
+			// if its the end of data. finish decrypting
+			if iterator >= stat.Size() {
+				break
+			}
+			n, err := file.ReadAt(iv[12:], iterator)
+			if err != nil {
```
[the remark this answers is not in the document]
Comment (joshua-goldstein) [reply ee/audit/run_ee.go:238]: done - added break statements

File: ee/audit/run_ee.go (hunk 15 above)
Comment (joshua-goldstein) [reply ee/audit/run_ee.go:198]: yes, good catch - added break statement here

**Comment by joshua-goldstein** [comment]
> I think we should consider writing more robust functions and test. This needs more work.

I have already added 4 integration tests, and we have a unit test to check basic parsing.  What additional tests are necessary?

File: ee/audit/run_ee.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (joshua-goldstein) [reply ee/audit/run_ee.go:113]: Additionally the functions `decryptBody` and `decryptBodyDeprecated` have very different parsing logic (the way they increment and get the length of text is fundamentally different).  They should not be written in one function.

File: ee/audit/run_ee.go [hunk 23]
```diff
@@ -107,37 +107,161 @@ func run() error {
...
 
-	iv := make([]byte, aes.BlockSize)
-	x.Check2(file.ReadAt(iv, iterator))
-	iterator = iterator + aes.BlockSize
+	// decrypt header in audit log to verify encryption key
+	// [16]byte IV + [4]byte len(x.VerificationText) + [11]byte x.VerificationText
+	decryptHeader := func() ([]byte, int64, error) {
+		var iterator int64 = 0
+		iv := make([]byte, aes.BlockSize)
+		_, err := file.ReadAt(iv, iterator) // get first iv
```
[the remark this answers is not in the document]
Comment (joshua-goldstein) [reply ee/audit/run_ee.go]: we now return the number of bits written `n` and use this to increment the iterator

File: ee/audit/run_ee.go [hunk 24]
```diff
@@ -107,37 +107,161 @@ func run() error {
...
+		if err != nil {
+			return nil, 0, err
+		}
+		iterator = iterator + aes.BlockSize + 4 // length of verification text encoded in uint32
 
-	t := make([]byte, len(x.VerificationText))
-	x.Check2(file.ReadAt(t, iterator))
-	iterator = iterator + int64(len(x.VerificationText))
+		ct := make([]byte, len(x.VerificationText))
+		_, err = file.ReadAt(ct, iterator)
```
[the remark this answers is not in the document]
Comment (joshua-goldstein) [reply ee/audit/run_ee.go]: we now return the number of bits written `n` and use this to increment the iterator

File: ee/audit/run_ee.go [hunk 25]
```diff
@@ -107,37 +107,161 @@ func run() error {
...
-	for {
-		// if its the end of data. finish decrypting
-		if iterator >= stat.Size() {
-			break
+	// [12]byte baseIV + [4]byte len(x.VerificationTextDeprecated) + [11]byte x.VerificationTextDeprecated
+	decryptHeaderDeprecated := func() ([]byte, int64, error) {
+		var iterator int64 = 0
+
+		iv := make([]byte, aes.BlockSize)
+		_, err := file.ReadAt(iv, iterator)
```
[the remark this answers is not in the document]
Comment (joshua-goldstein) [reply ee/audit/run_ee.go]: we now return the number of bits written `n` and use this to increment the iterator

File: ee/audit/run_ee.go [hunk 26]
```diff
@@ -107,37 +107,161 @@ func run() error {
...
 		}
-		x.Check2(file.ReadAt(iv[12:], iterator))
-		iterator = iterator + 4
+		iterator = iterator + aes.BlockSize
 
-		content := make([]byte, binary.BigEndian.Uint32(iv[12:]))
-		x.Check2(file.ReadAt(content, iterator))
-		iterator = iterator + int64(binary.BigEndian.Uint32(iv[12:]))
+		ct := make([]byte, len(x.VerificationTextDeprecated))
+		_, err = file.ReadAt(ct, iterator)
```
[the remark this answers is not in the document]
Comment (joshua-goldstein) [reply ee/audit/run_ee.go]: we now return the number of bytes written `n` and use this to increment the iterator

**Comment by joshua-goldstein** [comment]
> I didn't realize that byte array/slice can't be declared as a constant. This article has a few ideas https://blog.boot.dev/golang/golang-constant-maps-slices/. I don't mind if we declare a function that returns the constant value. I think it is useful to avoid a allocation that converts byte slice to string.

What is the purpose of this?  We are doing exactly one allocation per log file (default ~200 mb), so this is not a memory issue?

File: ee/audit/run_ee.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (joshua-goldstein) [reply ee/audit/run_ee.go:113]: As discussed we will refactor to use io.Reader object

File: ee/audit/run_ee.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (joshua-goldstein) [reply ee/audit/run_ee.go:118]: todo: `error.Wrap(err, "unable to read IV")` e.g.

File: ee/audit/run_ee.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (joshua-goldstein) [reply ee/audit/run_ee.go:126]: Refactored to use `file io.ReaderAt, outfile io.Writer`

File: ee/audit/run_ee.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (joshua-goldstein) [reply ee/audit/run_ee.go:131]: added error wrapper

File: ee/audit/run_ee.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (joshua-goldstein) [reply ee/audit/run_ee.go:138]: added error wrapper

File: ee/audit/run_ee.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (joshua-goldstein) [reply ee/audit/run_ee.go:164]: added error wrapper

File: testutil/testaudit/audit.go [hunk 27]
```diff
@@ -0,0 +1,60 @@
...
+	}
+	for _, m := range cmds {
+		if !logMap[m] {
+			t.Fatalf("audit logs not present for command %s", m)
+		}
+	}
+}
+
+// run manually to generate the encrypted audit log used in TestDecrypt
+func TestGenerateAuditForTestDecrypt(t *testing.T) {
```
Comment (joshua-goldstein) [inline testutil/testaudit/audit.go:42]: This is here so that in the future it is easy to generate an encrypted audit log; of course we only want to run this manually, so we should skip it

**Comment by joshua-goldstein** [comment]
The scope of this PR is now much larger than originally intended.  This PR now introduces:
- Four new integration tests for encrypted audit logging in `systest/audit_encrypted/audit_test.go`, two of which verify that encrypted audit logs can be encrypted, and two that verify that deprecated audit logs can be decrypted
- One new unit test in `ee/audit/run_ee_test.go` to verify that truncating an encrypted audit log does not cause the decrypt process to panic
- A modified unit test to decrypt using the new audit log format

This PR also does some minor refactoring to the code.

File: testutil/testaudit/audit.go (hunk 27 above)
Comment (mangalaman93) [reply testutil/testaudit/audit.go:42]: can we mention this in the code itself.

File: testutil/testaudit/docker-compose.yml [hunk 28]
```diff
@@ -0,0 +1,46 @@
...
+    working_dir: /data/alpha1
+    ports:
+      - "8080"
+      - "9080"
+    volumes:
+      - type: bind
+        source: $GOPATH/bin
+        target: /gobin
+        read_only: true
+      - type: bind
```
Comment (mangalaman93) [inline testutil/testaudit/docker-compose.yml:14]: I still don't remember why can't we use volume instead of this

File: x/log_writer.go [hunk 29]
```diff
@@ -41,7 +39,9 @@ const (
 	backupTimeFormat = "2006-01-02T15-04-05.000"
 	bufferSize       = 256 * 1024
 	flushInterval    = 10 * time.Second
-	VerificationText = "Hello World"
+	//  old logs before https://github.com/dgraph-io/dgraph/pull/8323 contain deprecated verification text in header
+	VerificationTextDeprecated = "Hello World"
```
Comment (mangalaman93) [inline x/log_writer.go:43]: would it be more useful to version them or something? What if we need to change this tomorrow again? We can also just do it the next time if that makes more sense.

**Review by mangalaman93 — CHANGES_REQUESTED** [review]
Did a quick review for now.  I will review it again tomorrow.

File: testutil/testaudit/audit.go (hunk 27 above)
Comment (joshua-goldstein) [reply testutil/testaudit/audit.go:42]: I already have a comment above the test - what else do you suggest?

File: ee/audit/run_ee.go [hunk 30]
```diff
@@ -107,39 +108,175 @@ func run() error {
...
-	var iterator int64 = 0
 
-	iv := make([]byte, aes.BlockSize)
-	x.Check2(file.ReadAt(iv, iterator))
-	iterator = iterator + aes.BlockSize
+	if err := decrypt(file, outfile, block, stat.Size()); err != nil {
+		return errors.Wrap(err, "could not decrypt audit log")
+	}
+
+	glog.Infof("Decryption of Audit file %s is Done. Decrypted file is %s",
```
[the remark this answers is not in the document]
Comment (joshua-goldstein) [reply ee/audit/run_ee.go:116]: made it lowercase and shorter

File: ee/audit/run_ee.go [hunk 31]
```diff
@@ -107,39 +108,175 @@ func run() error {
...
+		decryptCmd.Conf.GetString("in"),
+		decryptCmd.Conf.GetString("out"))
+	return nil
+}
+
+func decrypt(file io.ReaderAt, outfile io.Writer, block cipher.Block, sz int64) error {
+
```
[the remark this answers is not in the document]
Comment (joshua-goldstein) [reply ee/audit/run_ee.go:123]: done

File: ee/audit/run_ee.go [hunk 32]
```diff
@@ -107,39 +108,175 @@ func run() error {
...
+	// #################################################################
+	decryptBody := func() {
+		for {
+			// if its the end of data. finish decrypting
+			if iterator >= sz {
+				break
+			}
+			n, err := file.ReadAt(iv, iterator)
+			if err != nil {
+				glog.Warningf("received %v while decrypting audit log\n", err)
```
[the remark this answers is not in the document]
Comment (joshua-goldstein) [reply ee/audit/run_ee.go:202]: removed

File: ee/audit/run_ee.go [hunk 33]
```diff
@@ -107,39 +108,175 @@ func run() error {
...
+				glog.Warningf("read %v bytes, expected %v\n", n, len(iv))
+				break
+			}
+			iterator = iterator + 16
+			length := make([]byte, 4)
+			n, err = file.ReadAt(length, iterator)
+			if err != nil {
```
[the remark this answers is not in the document]
Comment (joshua-goldstein) [reply ee/audit/run_ee.go:209]: removed

File: ee/audit/run_ee_test.go [hunk 34]
```diff
@@ -0,0 +1,117 @@
...
+		case 540 <= i && i <= 790:
+			// at this point the output file will be empty
+			check(t, file.Truncate(sz-int64(i)))
+			// verify that decrypt does not panic
+			check(t, decrypt(file, outfile, block, sz))
+			check(t, outfile.Truncate(0))
+			_, err := outfile.Seek(0, 0)
+			check(t, err)
+		}
+
```
[the remark this answers is not in the document]
Comment (joshua-goldstein) [reply ee/audit/run_ee_test.go:115]: done

File: x/log_writer.go (hunk 29 above)
Comment (joshua-goldstein) [reply x/log_writer.go:43]: I hope we never have to look at this ever again :)  But in all seriousness I don't think we need to version it at this point.  If we do get in that position tomorrow we can just change these variable names.

File: testutil/testaudit/docker-compose.yml (hunk 28 above)
Comment (joshua-goldstein) [reply testutil/testaudit/docker-compose.yml:14]: I tried using a volume and ran into issues.  Then we tried debugging the problem together and couldn't figure it out.  The problem resolved when we switched back to using a bind mount.  We use bind mounts in the other audit log tests (that existed in the repo previously in `systest/audit`).  I don't want to introduce that kind of change here in this PR, because if we change it here we should also change it elsewhere too.

File: ee/audit/run_ee_test.go [hunk 35]
```diff
@@ -0,0 +1,117 @@
...
+		}
+	}()
+	defer os.RemoveAll("testfiles/zero_audit_0_1.log")
+	block, err := aes.NewCipher(key)
+	check(t, err)
+
+	zeroCmd := []string{"/removeNode", "/assign", "/moveTablet"}
+
+	// size of audit log is 825 bytes
+	for i := 0; i <= int(sz); i = i + 5 {
```
[the remark this answers is not in the document]
Comment (joshua-goldstein) [reply ee/audit/run_ee_test.go:83]: added a comment

File: ee/audit/run_ee_test.go [hunk 36]
```diff
@@ -0,0 +1,117 @@
...
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+// we will truncate copy of encrypted audit log file
+func copy(t *testing.T, src string, dst string) {
+	// could also us io.CopyN but this is a small file
+	data, err := os.ReadFile(src)
+	check(t, err)
```
[the remark this answers is not in the document]
Comment (joshua-goldstein) [reply ee/audit/run_ee_test.go:37]: thanks, I thought there should be a better way...  using testify now

File: ee/zero_audit_0_1.log.enc
Comment (mangalaman93) [inline ee/zero_audit_0_1.log.enc:1]: is this file needed?

**Review by mangalaman93 — APPROVED** [review]

File: ee/zero_audit_0_1.log.enc
Comment (joshua-goldstein) [reply ee/zero_audit_0_1.log.enc:1]: nope, good catch...  removing

**Review by mangalaman93 — APPROVED** [review]

**Review by skrdgraph — APPROVED** [review]
